### PR TITLE
Clone crash during s3

### PR DESCRIPTION
**Pull request by warniaha**
Added copy constructor for DaosBucket

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/rgw/rgw_sal_daos.h [hunk 1]
```diff
@@ -267,7 +272,7 @@ class DaosBucket : public Bucket {
                               uint64_t timeout) override;
   virtual int purge_instance(const DoutPrefixProvider* dpp) override;
   virtual std::unique_ptr<Bucket> clone() override {
-    return std::make_unique<DaosBucket>(*this);
+    return std::unique_ptr<DaosBucket>(new DaosBucket(*this));
```
Comment (zalsader) [inline src/rgw/rgw_sal_daos.h:275]: This is what `make_unique` literally does, so no new changes here.

File: src/rgw/rgw_sal_daos.h [hunk 2]
```diff
@@ -185,6 +185,11 @@ class DaosBucket : public Bucket {
 
   DaosBucket(DaosStore* _st) : store(_st), acls() {}
 
+  DaosBucket(const DaosBucket& _daos_bucket) : store(_daos_bucket.store), acls(), coh(DAOS_HDL_INVAL), dfs(nullptr) {
+    //TODO: deep copy all objects
+    //daos_duplicate_handle(_daos_bucket.coh);
+  }
+
```
Comment (zalsader) [inline src/rgw/rgw_sal_daos.h:192]: Maybe we could get rid of the copy constructor and just zero out the `coh` and `dfs` handles inside the `clone` method? This way we don't have to handle all the cases.

File: src/rgw/rgw_sal_daos.cc [hunk 3]
```diff
@@ -1254,6 +1254,7 @@ int DaosObject::get_obj_state(const DoutPrefixProvider* dpp,
                      << dendl;
   etag_bl.append(etag);
   state.attrset[RGW_ATTR_ETAG] = etag_bl;
+  *_state = &state;
```
Comment (zalsader) [inline src/rgw/rgw_sal_daos.cc:1257]: This was forgotten, thank you for including this.

File: src/rgw/rgw_sal_daos.cc [hunk 4]
```diff
@@ -2866,7 +2867,12 @@ std::unique_ptr<Object> DaosStore::get_object(const rgw_obj_key& k) {
 int DaosStore::get_bucket(const DoutPrefixProvider* dpp, User* u,
                           const rgw_bucket& b, std::unique_ptr<Bucket>* bucket,
                           optional_yield y) {
-  ldpp_dout(dpp, 20) << "DEBUG: get_bucket1: User" << u->get_id() << dendl;
+  if (u) {
+    ldpp_dout(dpp, 20) << "DEBUG: get_bucket1: User: " << u->get_id() << dendl;
+  }
+  else {
+    ldpp_dout(dpp, 20) << "DEBUG: get_bucket1: User: (nullptr)" << dendl;
+  }
```
Comment (zalsader) [inline src/rgw/rgw_sal_daos.cc:2875]: A ternary operator could make this less redundant: `(u? u->get_id(): "nullptr")`

**Review by zalsader — COMMENTED** [review]
Looks good, a bunch of small comments

File: src/rgw/rgw_sal_daos.h (hunk 1 above)
Comment (warniaha) [reply src/rgw/rgw_sal_daos.h:275]: I know that is the expectation, but there seems to be a subtle difference and the change does resolve a crash.  The copy constructor is also needed.

File: src/rgw/rgw_sal_daos.h (hunk 2 above)
Comment (warniaha) [reply src/rgw/rgw_sal_daos.h:192]: its part of the crash fix

File: src/rgw/rgw_sal_daos.cc (hunk 4 above)
Comment (warniaha) [reply src/rgw/rgw_sal_daos.cc:2875]: the operation u-get_id() doesn't return a string, so you can't

Comment (warniaha) [reply src/rgw/rgw_sal_daos.cc:2875]: I could add and operator<< that takes a User* which would improve the readability

**Comment by zalsader** [comment]
Could you please provide steps to reproduce the crash?

**Review by zalsader — APPROVED** [review]
Looks good!

**Pull request merged**